pkg/kube/informer.go

```diff
@@ -0,0 +1,295 @@
+
+// AddEventHandler adds the handler to each namespaced informer.
+func (i *multiNamespaceInformer) AddEventHandler(handler cache.ResourceEventHandler) {
+	for _, informer := range i.namespaceToInformer {
+		informer.AddEventHandler(handler)
+	}
+}
+
+// AddEventHandlerWithResyncPeriod adds the handler with a resync period to each
+// namespaced informer.
+func (i *multiNamespaceInformer) AddEventHandlerWithResyncPeriod(handler cache.ResourceEventHandler, resyncPeriod time.Duration) {
+	for _, informer := range i.namespaceToInformer {
+		informer.AddEventHandlerWithResyncPeriod(handler, resyncPeriod)
+	}
+}
+
+// AddIndexers adds the indexer for each namespaced informer.
+func (i *multiNamespaceInformer) AddIndexers(indexers cache.Indexers) error {
+	for _, informer := range i.namespaceToInformer {
+		err := informer.AddIndexers(indexers)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// HasSynced checks if each namespaced informer has synced.
+func (i *multiNamespaceInformer) HasSynced() bool {
+	for _, informer := range i.namespaceToInformer {
+		if ok := informer.HasSynced(); !ok {
+			return ok
+		}
+	}
+	return true
+}
+
+// Get fetches the object names by the given key, and places it in out.
+func (i *multiNamespaceInformer) Get(key types.NamespacedName, out runtime.Object) error {
+	indexer, err := i.indexerForNamespace(key.Namespace)
+	if err != nil {
+		return err
+	}
+
+	obj, exists, err := indexer.GetByKey(key.String())
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		return kerrors.NewNotFound(i.resource.GroupResource(), key.String())
+	}
+
+	outVal := reflect.ValueOf(out)
+	objVal := reflect.ValueOf(obj)
+	if !objVal.Type().AssignableTo(outVal.Type()) {
+		return fmt.Errorf("cache had type %s, but %s was asked for", objVal.Type(), outVal.Type())
+	}
+	reflect.Indirect(outVal).Set(reflect.Indirect(objVal))
+
+	// TODO(bison): Set GroupVersionKind?
+	// out.GetObjectKind().SetGroupVersionKind(groupVersionKind)
+
+	return nil
+}
+
+type ListOptions struct {
+	namespace string
+	selector  labels.Selector
+}
+
+func DefaultListOptions() *ListOptions {
+	return &ListOptions{
+		selector:  labels.Everything(),
+		namespace: metav1.NamespaceAll,
+	}
+}
+
+func (l *ListOptions) ApplyOptions(opts ...ListOption) *ListOptions {
+	for _, opt := range opts {
+		opt.ApplyToList(l)
+	}
+
+	return l
+}
+
+type ListOption interface {
+	ApplyToList(*ListOptions)
+}
+
+type InNamespace string
+
+func (n InNamespace) ApplyToList(opts *ListOptions) {
+	opts.namespace = string(n)
+}
+
+type MatchingLabelsSelector struct {
+	labels.Selector
+}
+
+func (m MatchingLabelsSelector) ApplyToList(opts *ListOptions) {
+	opts.selector = m
+}
+
+// List will list all objects in the informers' caches, across all watched
+// namespaces, matching the given label selector and set them in the Items field
+// of the provided list object.
+func (i *multiNamespaceInformer) List(out runtime.Object, opts ...ListOption) error {
```

Comment:
Why do we need our own ListOption?

Reply:
As opposed to using `ListOptions` from client-go? This is basically another place that is just mimicking / copying the controller-runtime API. Instead of trying to support the full client-go `ListOptions`, which I'm not sure is possible, this just adds a few helpers for limiting by namespace, or passing a label selector.

I definitely don't think this API is great. This bit is basically my main concern with this experiment. I'm pretty close to being convinced it's better to just generate the typed listers we need in a way that works with multiple indexes.